_posts/2019-08-18-Securing-Web-Applications-With-Keycloak.md

```diff
@@ -0,0 +1,608 @@
+
+## Creating a Client
+
+Every application that interacts with Keycloak is considered to be a client.  
+Let's create one for the [Single Page App](https://en.wikipedia.org/wiki/Single-page_application) or SPA.  
+
+Look for the `Clients` tab in the menu and hit `Create`.  
+Pick a name you think is suitable and choose OpenID Connect (OIDC) as protocol.  
+The `Root URL` can remain blank.  
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/create_client.png" width="50%">
+</div>
+
+After saving, we can see all the configuration options of the client.  
+* `Valid Redirect URIs` should be set to `http://localhost:4200/*` as this is the address of our SPA  
+* Allow all origins for testing purposes
+* Single Page apps use a `public client` because they can not securely hide client credentials
+* Make sure the `Standard Flow` is enabled. All other flows can be disabled
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/configure_client.png" width="70%">
+</div>
+
+> Standard flow is another name for the [Authorization Code Flow](https://tools.ietf.org/html/rfc6749#section-1.3.1) as defined in [the OAuth 2.0 specification](https://tools.ietf.org/html/rfc6749).
+
+> `Direct Access Grants Enabled` may remain enabled for now. It will be easy to test our configuration later.
+
+Don't forget to hit `Save` at the bottom of the form!
+
+## Creating Roles and Scopes
+Roles and scopes can be used to provide fine-grained access control to resources. 
+We want them to be present when handling requests with our Spring Boot application. 
+This part is optional, but can provide better insight in managing access to resources down the road.
+<br>
+Under `Roles` > `Add Role`, enter any name you like. Some standard roles are typically user or admin.
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/create_role.png" width="80%">
+</div>
+
+A scope is created in a similar way under `Client Scopes` > `Create`.  
+We will not show a consent screen, so you can uncheck this option.
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/create_scope.png" width="80%">
+</div>
+
+Now add the scope to the client we created earlier under `Clients` > `Client Scopes`.
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/assign_scope.png" width="80%">
+</div>
+<br>
+
+## Creating a user
+
+To use our application later, we need a user to log in with.
+In most corporate environments, users are stored in Active Directory or LDAP, which Keycloak can be connected to.
+For our example, however, we will simply create a user in Keycloak itself.  
+
+Search for the `Users` tab in the menu on the left and click `Add User`.  
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/create_user.png" width="60%">
```

Comment:
Add an image with eg. `<img class="image fit" src="{{ '/img/devoxx-ma-2018/speakers-dinner.jpg' | prepend: site.baseurl }}" alt="Speakers Dinner" width="60%" />`, otherwise this won't scale good.